File: converters/python/yalf2sqlite.py
#!/usr/bin/env python3
"""
YALF to SQLite Converter

Converts YALF (YALF Ain't Localisation Format) dictionary files to SQLite database.

Author: AnmiTaliDev <anmitalidev@nuros.org>
License: MIT
"""

import sqlite3
import yaml
import json
import argparse
import sys
from pathlib import Path
from typing import List, Dict, Any, Optional


class YALFConverter:
    """Converter from YALF format to SQLite database."""

    def __init__(self, db_path: str):
        """Initialize converter with database path."""
        self.db_path = db_path
        self.conn: Optional[sqlite3.Connection] = None

    def create_schema(self):
        """Create database schema for YALF data."""
        cursor = self.conn.cursor()

        # Main entries table
        cursor.execute("""
            CREATE TABLE IF NOT EXISTS entries (
                id INTEGER PRIMARY KEY,
                word TEXT NOT NULL,
                parent_id INTEGER,
                type TEXT NOT NULL,
                transcription TEXT,
                root_word TEXT,
                etymology TEXT,
                history TEXT,
                FOREIGN KEY (parent_id) REFERENCES entries(id)
            )
        """)

        # Writing systems table (flexible key-value storage)
        cursor.execute("""
            CREATE TABLE IF NOT EXISTS writing_systems (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                entry_id INTEGER NOT NULL,
                system_name TEXT NOT NULL,
                representation TEXT NOT NULL,
                FOREIGN KEY (entry_id) REFERENCES entries(id) ON DELETE CASCADE,
                UNIQUE(entry_id, system_name)
            )
        """)

        # Definitions table
        cursor.execute("""
            CREATE TABLE IF NOT EXISTS definitions (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                entry_id INTEGER NOT NULL,
                meaning TEXT NOT NULL,
                position INTEGER NOT NULL,
                FOREIGN KEY (entry_id) REFERENCES entries(id) ON DELETE CASCADE
            )
        """)

        # Translations table (flexible for any language)
        cursor.execute("""
            CREATE TABLE IF NOT EXISTS translations (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                definition_id INTEGER NOT NULL,
                language_code TEXT NOT NULL,
                translation TEXT NOT NULL,
                FOREIGN KEY (definition_id) REFERENCES definitions(id) ON DELETE CASCADE
            )
        """)

        # Examples table
        cursor.execute("""
            CREATE TABLE IF NOT EXISTS examples (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                definition_id INTEGER NOT NULL,
                position INTEGER NOT NULL,
                FOREIGN KEY (definition_id) REFERENCES definitions(id) ON DELETE CASCADE
            )
        """)

        # Example translations table
        cursor.execute("""
            CREATE TABLE IF NOT EXISTS example_translations (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                example_id INTEGER NOT NULL,
                language_code TEXT NOT NULL,
                text TEXT NOT NULL,
                FOREIGN KEY (example_id) REFERENCES examples(id) ON DELETE CASCADE
            )
        """)

        # Synonyms table
        cursor.execute("""
            CREATE TABLE IF NOT EXISTS synonyms (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                entry_id INTEGER NOT NULL,
                word_id INTEGER,
                word TEXT NOT NULL,
                note TEXT,
                FOREIGN KEY (entry_id) REFERENCES entries(id) ON DELETE CASCADE,
                FOREIGN KEY (word_id) REFERENCES entries(id)
            )
        """)

        # Create indexes for better query performance
        cursor.execute("CREATE INDEX IF NOT EXISTS idx_entries_word ON entries(word)")
        cursor.execute("CREATE INDEX IF NOT EXISTS idx_entries_type ON entries(type)")
        cursor.execute("CREATE INDEX IF NOT EXISTS idx_entries_parent ON entries(parent_id)")
        cursor.execute("CREATE INDEX IF NOT EXISTS idx_writing_systems_entry ON writing_systems(entry_id)")
        cursor.execute("CREATE INDEX IF NOT EXISTS idx_definitions_entry ON definitions(entry_id)")
        cursor.execute("CREATE INDEX IF NOT EXISTS idx_synonyms_entry ON synonyms(entry_id)")

        self.conn.commit()

    def insert_entry(self, entry: Dict[str, Any]):
        """Insert a single YALF entry into the database."""
        cursor = self.conn.cursor()

        # Insert main entry
        cursor.execute("""
            INSERT INTO entries (id, word, parent_id, type, transcription, root_word, etymology, history)
            VALUES (?, ?, ?, ?, ?, ?, ?, ?)
        """, (
            entry['id'],
            entry['word'],
            entry.get('parent_id'),
            entry['type'],
            entry.get('transcription'),
            entry.get('root_word'),
            entry.get('etymology'),
            entry.get('history')
        ))

        entry_id = entry['id']

        # Insert writing systems
        if 'writing_systems' in entry:
            for system_name, representation in entry['writing_systems'].items():
                cursor.execute("""
                    INSERT INTO writing_systems (entry_id, system_name, representation)
                    VALUES (?, ?, ?)
                """, (entry_id, system_name, representation))

        # Insert definitions
        if 'definitions' in entry:
            for pos, definition in enumerate(entry['definitions']):
                cursor.execute("""
                    INSERT INTO definitions (entry_id, meaning, position)
                    VALUES (?, ?, ?)
                """, (entry_id, definition['meaning'], pos))

                definition_id = cursor.lastrowid

                # Insert translations
                for key, value in definition.items():
                    if key.startswith('translation_'):
                        lang_code = key.replace('translation_', '')
                        cursor.execute("""
                            INSERT INTO translations (definition_id, language_code, translation)
                            VALUES (?, ?, ?)
                        """, (definition_id, lang_code, value))

                # Insert examples
                if 'examples' in definition:
                    for ex_pos, example in enumerate(definition['examples']):
                        cursor.execute("""
                            INSERT INTO examples (definition_id, position)
                            VALUES (?, ?)
                        """, (definition_id, ex_pos))

                        example_id = cursor.lastrowid

                        # Insert example translations
                        for lang_code, text in example.items():
                            cursor.execute("""
                                INSERT INTO example_translations (example_id, language_code, text)
                                VALUES (?, ?, ?)
                            """, (example_id, lang_code, text))

        # Insert synonyms
        if 'synonyms' in entry:
            for synonym in entry['synonyms']:
                cursor.execute("""
                    INSERT INTO synonyms (entry_id, word_id, word, note)
                    VALUES (?, ?, ?, ?)
                """, (entry_id, synonym.get('word_id'), synonym['word'], synonym.get('note')))

        self.conn.commit()

    def convert(self, yalf_path: str, overwrite: bool = False):
        """Convert YALF file to SQLite database."""
        # Check if database exists
        if Path(self.db_path).exists() and not overwrite:
            print(f"Error: Database {self.db_path} already exists. Use --overwrite to replace it.")
            return False

        # Remove existing database if overwrite is True
        if overwrite and Path(self.db_path).exists():
            Path(self.db_path).unlink()

        # Read YALF file
        try:
            with open(yalf_path, 'r', encoding='utf-8') as f:
                entries = yaml.safe_load(f)
        except Exception as e:
            print(f"Error reading YALF file: {e}")
            return False

        if not isinstance(entries, list):
            print("Error: YALF file must contain a list of entries")
            return False

        # Connect to database
        try:
            self.conn = sqlite3.connect(self.db_path)
            self.create_schema()
        except Exception as e:
            print(f"Error creating database: {e}")
            return False

        # Insert entries
        try:
            for entry in entries:
                self.insert_entry(entry)
            print(f"Successfully converted {len(entries)} entries to {self.db_path}")
            return True
        except Exception as e:
            print(f"Error inserting entries: {e}")
            self.conn.rollback()
            return False
        finally:
            if self.conn:
                self.conn.close()


def main():
    """Main entry point for the converter."""
    parser = argparse.ArgumentParser(
        description='Convert YALF dictionary files to SQLite database',
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
Examples:
  %(prog)s dictionary.yalf output.db
  %(prog)s input.yaml output.db --overwrite
        """
    )

    parser.add_argument('input', help='Input YALF file (.yalf or .yaml)')
    parser.add_argument('output', help='Output SQLite database file (.db or .sqlite)')
    parser.add_argument('--overwrite', '-f', action='store_true',
                       help='Overwrite output database if it exists')
    parser.add_argument('--version', action='version', version='%(prog)s 1.0.0')

    args = parser.parse_args()

    # Check input file exists
    if not Path(args.input).exists():
        print(f"Error: Input file {args.input} not found")
        sys.exit(1)

    # Convert
    converter = YALFConverter(args.output)
    success = converter.convert(args.input, args.overwrite)

    sys.exit(0 if success else 1)


if __name__ == '__main__':
    main()

File: converters/go/yalf2sqlite.go
package main

import (
	"database/sql"
	"flag"
	"fmt"
	"os"
	"path/filepath"

	_ "github.com/mattn/go-sqlite3"
	"gopkg.in/yaml.v3"
)

// Entry represents a YALF dictionary entry
type Entry struct {
	ID             int               `yaml:"id"`
	Word           string            `yaml:"word"`
	ParentID       *int              `yaml:"parent_id"`
	Type           string            `yaml:"type"`
	Transcription  string            `yaml:"transcription,omitempty"`
	WritingSystems map[string]string `yaml:"writing_systems,omitempty"`
	RootWord       string            `yaml:"root_word,omitempty"`
	Etymology      string            `yaml:"etymology,omitempty"`
	History        string            `yaml:"history,omitempty"`
	Definitions    []Definition      `yaml:"definitions"`
	Synonyms       []Synonym         `yaml:"synonyms,omitempty"`
}

// Definition represents a word definition with translations
type Definition struct {
	Meaning  string                 `yaml:"meaning"`
	Examples []map[string]string    `yaml:"examples,omitempty"`
	Extra    map[string]interface{} `yaml:",inline"`
}

// Synonym represents a synonym entry
type Synonym struct {
	WordID *int   `yaml:"word_id"`
	Word   string `yaml:"word"`
	Note   string `yaml:"note,omitempty"`
}

// Converter handles YALF to SQLite conversion
type Converter struct {
	db *sql.DB
}

// NewConverter creates a new converter with database connection
func NewConverter(dbPath string) (*Converter, error) {
	db, err := sql.Open("sqlite3", dbPath)
	if err != nil {
		return nil, fmt.Errorf("failed to open database: %w", err)
	}
	return &Converter{db: db}, nil
}

// Close closes the database connection
func (c *Converter) Close() error {
	return c.db.Close()
}

// CreateSchema creates the database schema
func (c *Converter) CreateSchema() error {
	schema := `
	CREATE TABLE IF NOT EXISTS entries (
		id INTEGER PRIMARY KEY,
		word TEXT NOT NULL,
		parent_id INTEGER,
		type TEXT NOT NULL,
		transcription TEXT,
		root_word TEXT,
		etymology TEXT,
		history TEXT,
		FOREIGN KEY (parent_id) REFERENCES entries(id)
	);

	CREATE TABLE IF NOT EXISTS writing_systems (
		id INTEGER PRIMARY KEY AUTOINCREMENT,
		entry_id INTEGER NOT NULL,
		system_name TEXT NOT NULL,
		representation TEXT NOT NULL,
		FOREIGN KEY (entry_id) REFERENCES entries(id) ON DELETE CASCADE,
		UNIQUE(entry_id, system_name)
	);

	CREATE TABLE IF NOT EXISTS definitions (
		id INTEGER PRIMARY KEY AUTOINCREMENT,
		entry_id INTEGER NOT NULL,
		meaning TEXT NOT NULL,
		position INTEGER NOT NULL,
		FOREIGN KEY (entry_id) REFERENCES entries(id) ON DELETE CASCADE
	);

	CREATE TABLE IF NOT EXISTS translations (
		id INTEGER PRIMARY KEY AUTOINCREMENT,
		definition_id INTEGER NOT NULL,
		language_code TEXT NOT NULL,
		translation TEXT NOT NULL,
		FOREIGN KEY (definition_id) REFERENCES definitions(id) ON DELETE CASCADE
	);

	CREATE TABLE IF NOT EXISTS examples (
		id INTEGER PRIMARY KEY AUTOINCREMENT,
		definition_id INTEGER NOT NULL,
		position INTEGER NOT NULL,
		FOREIGN KEY (definition_id) REFERENCES definitions(id) ON DELETE CASCADE
	);

	CREATE TABLE IF NOT EXISTS example_translations (
		id INTEGER PRIMARY KEY AUTOINCREMENT,
		example_id INTEGER NOT NULL,
		language_code TEXT NOT NULL,
		text TEXT NOT NULL,
		FOREIGN KEY (example_id) REFERENCES examples(id) ON DELETE CASCADE
	);

	CREATE TABLE IF NOT EXISTS synonyms (
		id INTEGER PRIMARY KEY AUTOINCREMENT,
		entry_id INTEGER NOT NULL,
		word_id INTEGER,
		word TEXT NOT NULL,
		note TEXT,
		FOREIGN KEY (entry_id) REFERENCES entries(id) ON DELETE CASCADE,
		FOREIGN KEY (word_id) REFERENCES entries(id)
	);

	CREATE INDEX IF NOT EXISTS idx_entries_word ON entries(word);
	CREATE INDEX IF NOT EXISTS idx_entries_type ON entries(type);
	CREATE INDEX IF NOT EXISTS idx_entries_parent ON entries(parent_id);
	CREATE INDEX IF NOT EXISTS idx_writing_systems_entry ON writing_systems(entry_id);
	CREATE INDEX IF NOT EXISTS idx_definitions_entry ON definitions(entry_id);
	CREATE INDEX IF NOT EXISTS idx_synonyms_entry ON synonyms(entry_id);
	`

	_, err := c.db.Exec(schema)
	return err
}

// InsertEntry inserts a single YALF entry into the database
func (c *Converter) InsertEntry(entry Entry) error {
	tx, err := c.db.Begin()
	if err != nil {
		return err
	}
	defer tx.Rollback()

	// Insert main entry
	_, err = tx.Exec(`
		INSERT INTO entries (id, word, parent_id, type, transcription, root_word, etymology, history)
		VALUES (?, ?, ?, ?, ?, ?, ?, ?)
	`, entry.ID, entry.Word, entry.ParentID, entry.Type, entry.Transcription,
		entry.RootWord, entry.Etymology, entry.History)
	if err != nil {
		return fmt.Errorf("failed to insert entry: %w", err)
	}

	// Insert writing systems
	for systemName, representation := range entry.WritingSystems {
		_, err = tx.Exec(`
			INSERT INTO writing_systems (entry_id, system_name, representation)
			VALUES (?, ?, ?)
		`, entry.ID, systemName, representation)
		if err != nil {
			return fmt.Errorf("failed to insert writing system: %w", err)
		}
	}

	// Insert definitions
	for pos, definition := range entry.Definitions {
		result, err := tx.Exec(`
			INSERT INTO definitions (entry_id, meaning, position)
			VALUES (?, ?, ?)
		`, entry.ID, definition.Meaning, pos)
		if err != nil {
			return fmt.Errorf("failed to insert definition: %w", err)
		}

		definitionID, _ := result.LastInsertId()

		// Insert translations (from Extra field)
		for key, value := range definition.Extra {
			if len(key) > 12 && key[:12] == "translation_" {
				langCode := key[12:]
				if strVal, ok := value.(string); ok {
					_, err = tx.Exec(`
						INSERT INTO translations (definition_id, language_code, translation)
						VALUES (?, ?, ?)
					`, definitionID, langCode, strVal)
					if err != nil {
						return fmt.Errorf("failed to insert translation: %w", err)
					}
				}
			}
		}

		// Insert examples
		for exPos, example := range definition.Examples {
			result, err := tx.Exec(`
				INSERT INTO examples (definition_id, position)
				VALUES (?, ?)
			`, definitionID, exPos)
			if err != nil {
				return fmt.Errorf("failed to insert example: %w", err)
			}

			exampleID, _ := result.LastInsertId()

			// Insert example translations
			for langCode, text := range example {
				_, err = tx.Exec(`
					INSERT INTO example_translations (example_id, language_code, text)
					VALUES (?, ?, ?)
				`, exampleID, langCode, text)
				if err != nil {
					return fmt.Errorf("failed to insert example translation: %w", err)
				}
			}
		}
	}

	// Insert synonyms
	for _, synonym := range entry.Synonyms {
		_, err = tx.Exec(`
			INSERT INTO synonyms (entry_id, word_id, word, note)
			VALUES (?, ?, ?, ?)
		`, entry.ID, synonym.WordID, synonym.Word, synonym.Note)
		if err != nil {
			return fmt.Errorf("failed to insert synonym: %w", err)
		}
	}

	return tx.Commit()
}

// Convert converts a YALF file to SQLite database
func (c *Converter) Convert(yalfPath string) error {
	// Read YALF file
	data, err := os.ReadFile(yalfPath)
	if err != nil {
		return fmt.Errorf("failed to read YALF file: %w", err)
	}

	// Parse YAML
	var entries []Entry
	err = yaml.Unmarshal(data, &entries)
	if err != nil {
		return fmt.Errorf("failed to parse YAML: %w", err)
	}

	// Create schema
	err = c.CreateSchema()
	if err != nil {
		return fmt.Errorf("failed to create schema: %w", err)
	}

	// Insert entries
	for _, entry := range entries {
		err = c.InsertEntry(entry)
		if err != nil {
			return fmt.Errorf("failed to insert entry %d: %w", entry.ID, err)
		}
	}

	fmt.Printf("Successfully converted %d entries\n", len(entries))
	return nil
}

func main() {
	var (
		input     string
		output    string
		overwrite bool
		version   bool
	)

	flag.StringVar(&input, "input", "", "Input YALF file (.yalf or .yaml)")
	flag.StringVar(&input, "i", "", "Input YALF file (shorthand)")
	flag.StringVar(&output, "output", "", "Output SQLite database file (.db or .sqlite)")
	flag.StringVar(&output, "o", "", "Output SQLite database file (shorthand)")
	flag.BoolVar(&overwrite, "overwrite", false, "Overwrite output database if it exists")
	flag.BoolVar(&overwrite, "f", false, "Overwrite output database if it exists (shorthand)")
	flag.BoolVar(&version, "version", false, "Print version information")

	flag.Usage = func() {
		fmt.Fprintf(os.Stderr, "Usage: %s [OPTIONS] -i INPUT -o OUTPUT\n\n", filepath.Base(os.Args[0]))
		fmt.Fprintf(os.Stderr, "Convert YALF dictionary files to SQLite database\n\n")
		fmt.Fprintf(os.Stderr, "Options:\n")
		flag.PrintDefaults()
		fmt.Fprintf(os.Stderr, "\nExamples:\n")
		fmt.Fprintf(os.Stderr, "  %s -i dictionary.yalf -o output.db\n", filepath.Base(os.Args[0]))
		fmt.Fprintf(os.Stderr, "  %s -i input.yaml -o output.db -f\n", filepath.Base(os.Args[0]))
	}

	flag.Parse()

	if version {
		fmt.Println("yalf2sqlite version 1.0.0")
		fmt.Println("Author: AnmiTaliDev <anmitalidev@nuros.org>")
		os.Exit(0)
	}

	if input == "" || output == "" {
		flag.Usage()
		os.Exit(1)
	}

	// Check input file exists
	if _, err := os.Stat(input); os.IsNotExist(err) {
		fmt.Fprintf(os.Stderr, "Error: Input file %s not found\n", input)
		os.Exit(1)
	}

	// Check if output exists
	if _, err := os.Stat(output); err == nil && !overwrite {
		fmt.Fprintf(os.Stderr, "Error: Database %s already exists. Use -f to overwrite\n", output)
		os.Exit(1)
	}

	// Remove existing database if overwrite is true
	if overwrite {
		os.Remove(output)
	}

	// Create converter
	converter, err := NewConverter(output)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error: %v\n", err)
		os.Exit(1)
	}
	defer converter.Close()

	// Convert
	err = converter.Convert(input)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error: %v\n", err)
		os.Exit(1)
	}

	fmt.Printf("Successfully created database: %s\n", output)
}
